runtime/gc_structs/ClassLoaderClassesIterator.cpp

```diff
@@ -108,9 +108,25 @@ GC_ClassLoaderClassesIterator::nextClass()
 		if (ANONYMOUS_CLASSES == _mode) {
 			_nextClass = nextAnonymousClass();
 		} else {
-			if ( (result->classLoader == _classLoader) && (NULL != result->arrayClass) ) {
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+			if ((result->classLoader == _classLoader) && (NULL != result->nullRestrictedArrayClass)) {
+				_nullRestrictedArrayWalkState = TRUE;
+				_nextClass = result->nullRestrictedArrayClass;
+			} else
+#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
+			if ((result->classLoader == _classLoader) && (NULL != result->arrayClass)) {
 				/* this class is defined in the loader, so follow its array classes */
 				_nextClass = result->arrayClass;
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+			} else if (TRUE == _nullRestrictedArrayWalkState) {
+				J9Class *arrayClazz = ((J9ArrayClass *)result)->leafComponentType->arrayClass;
+				_nullRestrictedArrayWalkState = FALSE;
+				if ( (result->classLoader == _classLoader) && (NULL != arrayClazz)) {
+					_nextClass = arrayClazz;
+				} else {
+					_nextClass = nextTableClass();
```

Comment:
The only issue left is that the code assumes if `_nullRestrictedArrayWalkState` is true then `_mode == TABLE_CLASSES`. I think thats tru right now, but it may not be tru in the future.

Reply:
Skipping functional comments for now, style wise this is cpp code, not c and _nullRestrictedArrayWalkState is defined as bool. So, you should use true instead of TRUE, do if (_nullRestrictedArrayWalkState) instead of if (TRUE == _nullRestrictedArrayWalkState) etc. Also please add an initialization of the new variable in constructor.